File: Samples/ReadDecodedPixels/VB.NET/ReadDecodedPixels.vb
Imports System
Imports System.Diagnostics
Imports System.Drawing
Imports System.Drawing.Imaging
Imports System.Runtime.InteropServices
Imports BitMiracle.Jpeg2k

Module ReadDecodedPixels
    Sub Main()
        ' The library won't work without a license. You can get free time limited license
        ' at https://bitmiracle.com/jpeg2000/
        LicenseManager.SetTrialLicense("contact support@bitmiracle.com for a license")

        Dim fileName = "Sample data/p1_04.j2k"
        Using image = New J2kImage(fileName)
            Dim imageData As J2kImageData = image.Decode()

            Dim components = imageData.Components
            Dim comp0 = components(0)
            If comp0.BitsPerPixel < 8 Then
                Console.WriteLine("Unsupported bits per pixel value: {0}. At least 8 bpp required.", comp0.BitsPerPixel)
                Return
            End If

            Dim comp1 As J2kImageComponent = Nothing
            Dim comp2 As J2kImageComponent = Nothing
            If components.Count >= 3 Then
                comp1 = components(1)
                comp2 = components(2)
            End If

            Dim w = comp0.Width
            Dim h = comp0.Height
            If components.Count >= 3 AndAlso
                comp0.HorizontalSeparation = comp1.HorizontalSeparation AndAlso
                comp1.HorizontalSeparation = comp2.HorizontalSeparation AndAlso
                comp0.VerticalSeparation = comp1.VerticalSeparation AndAlso
                comp1.VerticalSeparation = comp2.VerticalSeparation AndAlso
                comp0.BitsPerPixel = comp1.BitsPerPixel AndAlso
                comp1.BitsPerPixel = comp2.BitsPerPixel Then

                ' 24 bpp
                Dim adjustR = 0
                If comp0.BitsPerPixel > 8 Then adjustR = comp0.BitsPerPixel - 8

                Dim adjustG = 0
                If comp1.BitsPerPixel > 8 Then adjustG = comp1.BitsPerPixel - 8

                Dim adjustB = 0
                If comp2.BitsPerPixel > 8 Then adjustB = comp2.BitsPerPixel - 8

                Dim redPixels As J2kPixels = comp0.GetPixels()
                Dim redAdjustment = If(comp0.PixelsAreSigned, 1 << comp0.BitsPerPixel - 1, 0)

                Dim greenPixels As J2kPixels = comp1.GetPixels()
                Dim greenAdjustment = If(comp1.PixelsAreSigned, 1 << comp1.BitsPerPixel - 1, 0)

                Dim bluePixels As J2kPixels = comp2.GetPixels()
                Dim blueAdjustment = If(comp2.PixelsAreSigned, 1 << comp2.BitsPerPixel - 1, 0)

                Using bitmap = New Bitmap(w, h, PixelFormat.Format24bppRgb)
                    Dim bd As BitmapData = bitmap.LockBits(New Rectangle(0, 0, w, h), ImageLockMode.WriteOnly, bitmap.PixelFormat)

                    Dim dest = New Byte(bd.Stride * bd.Height - 1) {}
                    Marshal.Copy(bd.Scan0, dest, 0, dest.Length)

                    Dim y = h - 1
                    While y >= 0
                        Dim start = y * w

                        Dim x = 0
                        While x < w
                            Dim r = redPixels(start + x) + redAdjustment
                            r = (r >> adjustR) + (r >> adjustR - 1) Mod 2
                            If r > 255 Then
                                r = 255
                            ElseIf r < 0 Then
                                r = 0
                            End If

                            Dim g = greenPixels(start + x) + greenAdjustment
                            g = (g >> adjustG) + (g >> adjustG - 1) Mod 2
                            If g > 255 Then
                                g = 255
                            ElseIf g < 0 Then
                                g = 0
                            End If

                            Dim b = bluePixels(start + x) + blueAdjustment
                            b = (b >> adjustB) + (b >> adjustB - 1) Mod 2
                            If b > 255 Then
                                b = 255
                            ElseIf b < 0 Then
                                b = 0
                            End If

                            Dim baseOffset = y * bd.Stride + x * 3
                            dest(baseOffset + 0) = CByte(b)
                            dest(baseOffset + 1) = CByte(g)
                            dest(baseOffset + 2) = CByte(r)

                            x += 1
                        End While

                        y -= 1
                    End While

                    Marshal.Copy(dest, 0, bd.Scan0, dest.Length)

                    bitmap.UnlockBits(bd)

                    Const OutputFileName = "result_24bpp.png"
                    bitmap.Save(OutputFileName, ImageFormat.Png)
                    Process.Start(OutputFileName)
                End Using
            Else
                ' 8 bpp grayscale
                Dim adjustG = 0
                If comp0.BitsPerPixel > 8 Then adjustG = comp0.BitsPerPixel - 8

                Dim rowPixels As J2kPixels = comp0.GetPixels()
                Using bitmap = New Bitmap(w, h, PixelFormat.Format8bppIndexed)
                    Dim pal = bitmap.Palette
                    For i = 0 To 255
                        pal.Entries(i) = Color.FromArgb(i, i, i)
                    Next

                    bitmap.Palette = pal

                    Dim bd As BitmapData = bitmap.LockBits(New Rectangle(0, 0, w, h), ImageLockMode.WriteOnly, bitmap.PixelFormat)

                    Dim dest = New Byte(bd.Stride * bd.Height - 1) {}
                    Marshal.Copy(bd.Scan0, dest, 0, dest.Length)

                    Dim y = h - 1
                    While y >= 0
                        Dim start = y * w

                        Dim x = 0
                        While x < w
                            Dim g = rowPixels(start + x) + adjustG
                            g = (g >> adjustG) + (g >> adjustG - 1) Mod 2
                            If g > 255 Then
                                g = 255
                            ElseIf g < 0 Then
                                g = 0
                            End If

                            dest(y * bd.Stride + x) = CByte(g)

                            x += 1
                        End While

                        y -= 1
                    End While

                    Marshal.Copy(dest, 0, bd.Scan0, dest.Length)

                    bitmap.UnlockBits(bd)

                    Const OutputFileName = "result_8bpp.png"
                    bitmap.Save(OutputFileName, ImageFormat.Png)
                    Process.Start(OutputFileName)
                End Using
            End If
        End Using
    End Sub
End Module

File: Samples/ReadDecodedPixels/C#/ReadDecodedPixels.cs
using System;
using System.Collections.ObjectModel;
using System.Diagnostics;
using System.Drawing;
using System.Drawing.Imaging;
using System.Runtime.InteropServices;
using BitMiracle.Jpeg2k;

namespace ReadDecodedPixels
{
    class ReadDecodedPixels
    {
        static void Main(string[] args)
        {
            // The library won't work without a license. You can get free time limited license
            // at https://bitmiracle.com/jpeg2000/
            LicenseManager.SetTrialLicense("contact support@bitmiracle.com for a license");

            string fileName = @"Sample data/p1_04.j2k";
            using (var image = new J2kImage(fileName))
            {
                J2kImageData imageData = image.Decode();

                ReadOnlyCollection<J2kImageComponent> components = imageData.Components;
                J2kImageComponent comp0 = components[0];
                if (comp0.BitsPerPixel < 8)
                {
                    Console.WriteLine("Unsupported bits per pixel value: {0}. At least 8 bpp required.", comp0.BitsPerPixel);
                    return;
                }

                J2kImageComponent comp1 = null;
                J2kImageComponent comp2 = null;
                if (components.Count >= 3)
                {
                    comp1 = components[1];
                    comp2 = components[2];
                }

                int w = comp0.Width;
                int h = comp0.Height;
                if (components.Count >= 3 &&
                    comp0.HorizontalSeparation == comp1.HorizontalSeparation &&
                    comp1.HorizontalSeparation == comp2.HorizontalSeparation &&
                    comp0.VerticalSeparation == comp1.VerticalSeparation &&
                    comp1.VerticalSeparation == comp2.VerticalSeparation &&
                    comp0.BitsPerPixel == comp1.BitsPerPixel &&
                    comp1.BitsPerPixel == comp2.BitsPerPixel)
                {
                    // 24 bpp
                    int adjustR = 0;
                    if (comp0.BitsPerPixel > 8)
                        adjustR = comp0.BitsPerPixel - 8;

                    int adjustG = 0;
                    if (comp1.BitsPerPixel > 8)
                        adjustG = comp1.BitsPerPixel - 8;

                    int adjustB = 0;
                    if (comp2.BitsPerPixel > 8)
                        adjustB = comp2.BitsPerPixel - 8;

                    J2kPixels redPixels = comp0.GetPixels();
                    int redAdjustment = comp0.PixelsAreSigned ? (1 << (comp0.BitsPerPixel - 1)) : 0;

                    J2kPixels greenPixels = comp1.GetPixels();
                    int greenAdjustment = comp1.PixelsAreSigned ? (1 << (comp1.BitsPerPixel - 1)) : 0;

                    J2kPixels bluePixels = comp2.GetPixels();
                    int blueAdjustment = comp2.PixelsAreSigned ? (1 << (comp2.BitsPerPixel - 1)) : 0;

                    using (var bitmap = new Bitmap(w, h, PixelFormat.Format24bppRgb))
                    {
                        BitmapData bd = bitmap.LockBits(new Rectangle(0, 0, w, h), ImageLockMode.WriteOnly, bitmap.PixelFormat);

                        byte[] dest = new byte[bd.Stride * bd.Height];
                        Marshal.Copy(bd.Scan0, dest, 0, dest.Length);

                        for (int y = h - 1; y >= 0; --y)
                        {
                            int start = y * w;
                            for (int x = 0; x < w; ++x)
                            {
                                int r = redPixels[start + x] + redAdjustment;
                                r = (r >> adjustR) + ((r >> (adjustR - 1)) % 2);
                                if (r > 255)
                                    r = 255;
                                else if (r < 0)
                                    r = 0;

                                int g = greenPixels[start + x] + greenAdjustment;
                                g = (g >> adjustG) + ((g >> (adjustG - 1)) % 2);
                                if (g > 255)
                                    g = 255;
                                else if (g < 0)
                                    g = 0;

                                int b = bluePixels[start + x] + blueAdjustment;
                                b = (b >> adjustB) + ((b >> (adjustB - 1)) % 2);
                                if (b > 255)
                                    b = 255;
                                else if (b < 0)
                                    b = 0;

                                int baseOffset = y * bd.Stride + x * 3;
                                dest[baseOffset + 0] = (byte)b;
                                dest[baseOffset + 1] = (byte)g;
                                dest[baseOffset + 2] = (byte)r;
                            }
                        }

                        Marshal.Copy(dest, 0, bd.Scan0, dest.Length);

                        bitmap.UnlockBits(bd);

                        const string OutputFileName = "result_24bpp.png";
                        bitmap.Save(OutputFileName, ImageFormat.Png);
                        Process.Start(OutputFileName);
                    }
                }
                else
                {
                    // 8 bpp grayscale
                    int adjustG = 0;
                    if (comp0.BitsPerPixel > 8)
                        adjustG = comp0.BitsPerPixel - 8;

                    J2kPixels rowPixels = comp0.GetPixels();
                    using (var bitmap = new Bitmap(w, h, PixelFormat.Format8bppIndexed))
                    {
                        ColorPalette pal = bitmap.Palette;
                        for (int i = 0; i <= 255; i++)
                            pal.Entries[i] = Color.FromArgb(i, i, i);

                        bitmap.Palette = pal;

                        BitmapData bd = bitmap.LockBits(new Rectangle(0, 0, w, h), ImageLockMode.WriteOnly, bitmap.PixelFormat);

                        byte[] dest = new byte[bd.Stride * bd.Height];
                        Marshal.Copy(bd.Scan0, dest, 0, dest.Length);

                        for (int y = h - 1; y >= 0; --y)
                        {
                            int start = y * w;
                            for (int x = 0; x < w; ++x)
                            {
                                int g = rowPixels[start + x] + adjustG;
                                g = (g >> adjustG) + ((g >> (adjustG - 1)) % 2);
                                if (g > 255)
                                    g = 255;
                                else if (g < 0)
                                    g = 0;

                                dest[y * bd.Stride + x] = (byte)g;
                            }
                        }

                        Marshal.Copy(dest, 0, bd.Scan0, dest.Length);

                        bitmap.UnlockBits(bd);

                        const string OutputFileName = "result_8bpp.png";
                        bitmap.Save(OutputFileName, ImageFormat.Png);
                        Process.Start(OutputFileName);
                    }
                }
            }
        }
    }
}
